File: backend/app/schemas/outcome.py
"""录用结果（效果回流）schema。

- ``OutcomeOut`` / ``OutcomeUpsertRequest``：结果录入与展示
- ``CalibrationReport``：评分 × 用人结果校准报告（评估报告 P1-5）
- ``FunnelStats``：招聘漏斗 + 渠道质量统计（评估报告 P1-4 最小闭环）
"""
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.models.outcome import FINAL_STATUS_VALUES

_FINAL_STATUS_PATTERN = r"^(hired|probation_passed|rejected|withdrawn)$"


# ============================================================================
# 结果录入
# ============================================================================


class OutcomeUpsertRequest(BaseModel):
    """HR 录入 / 更新最终用人结果。"""

    final_status: str = Field(pattern=_FINAL_STATUS_PATTERN)
    note: str | None = Field(default=None, max_length=2000)


class OutcomeOut(BaseModel):
    """结果行对外表示。"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    job_id: uuid.UUID
    candidate_id: uuid.UUID
    final_status: str
    note: str | None = None
    decided_at: datetime | None = None
    created_at: datetime | None = None


# ============================================================================
# 评分校准报告
# ============================================================================


class CalibrationBucket(BaseModel):
    """单个分数段的结果分布。"""

    score_min: int
    score_max: int
    total: int = 0
    hired: int = 0  # hired + probation_passed（正样本）
    rejected: int = 0  # rejected（负样本）
    other: int = 0  # withdrawn 等中性结果

    @computed_field  # type: ignore[prop-decorator]
    @property
    def hire_rate(self) -> float | None:
        """正样本命中率（分母 = 有明确结果的正负样本之和）。"""
        denom = self.hired + self.rejected
        if denom == 0:
            return None
        return round(self.hired / denom, 4)


class CalibrationReport(BaseModel):
    """评分 × 用人结果校准报告：分数越高 hire_rate 应单调递增。"""

    job_id: uuid.UUID | None = None
    buckets: list[CalibrationBucket]
    total_with_outcome: int = 0
    """已录入结果且完成评分的候选人数。"""


# ============================================================================
# 招聘漏斗
# ============================================================================


class ChannelQuality(BaseModel):
    """单渠道质量：来源 → 数量 → 通过率 → 录用数。"""

    source_type: str
    total: int = 0
    screened_pass: int = 0
    hired: int = 0

    @computed_field  # type: ignore[prop-decorator]
    @property
    def pass_rate(self) -> float | None:
        if self.total == 0:
            return None
        return round(self.screened_pass / self.total, 4)


class FunnelStats(BaseModel):
    """招聘漏斗统计（job 维度或 team 汇总）。"""

    job_id: uuid.UUID | None = None
    total_pool: int = 0  # 进入该 job 筛选池的候选人
    screened_pass: int = 0
    needs_review: int = 0
    disqualified: int = 0
    scored: int = 0
    interviewed: int = 0  # 存在面试会话
    hired: int = 0  # outcome ∈ {hired, probation_passed}
    channels: list[ChannelQuality] = Field(default_factory=list)


__all__ = [
    "OutcomeUpsertRequest",
    "OutcomeOut",
    "CalibrationBucket",
    "CalibrationReport",
    "ChannelQuality",
    "FunnelStats",
    "FINAL_STATUS_VALUES",
]

File: frontend/lib/api/outcome.ts
"use client";

import { apiClient } from "./client";

/**
 * 录用结果（效果回流）API：ground truth 落库 + 校准报告 + 漏斗统计。
 */

export type FinalStatus =
  | "hired"
  | "probation_passed"
  | "rejected"
  | "withdrawn";

export const FINAL_STATUS_LABELS: Record<FinalStatus, string> = {
  hired: "已录用",
  probation_passed: "已转正",
  rejected: "未录用",
  withdrawn: "候选人放弃",
};

export interface CandidateJobOutcome {
  id: string;
  job_id: string;
  candidate_id: string;
  final_status: FinalStatus;
  note: string | null;
  decided_at: string | null;
  created_at: string | null;
}

export interface CalibrationBucket {
  score_min: number;
  score_max: number;
  total: number;
  hired: number;
  rejected: number;
  other: number;
  hire_rate: number | null;
}

export interface CalibrationReport {
  job_id: string | null;
  buckets: CalibrationBucket[];
  total_with_outcome: number;
}

export interface ChannelQuality {
  source_type: string;
  total: number;
  screened_pass: number;
  hired: number;
  pass_rate: number | null;
}

export interface FunnelStats {
  job_id: string | null;
  total_pool: number;
  screened_pass: number;
  needs_review: number;
  disqualified: number;
  scored: number;
  interviewed: number;
  hired: number;
  channels: ChannelQuality[];
}

export async function putOutcomeApi(
  jobId: string,
  candidateId: string,
  payload: { final_status: FinalStatus; note?: string | null },
): Promise<CandidateJobOutcome> {
  const { data } = await apiClient.put<CandidateJobOutcome>(
    `/api/jobs/${jobId}/candidates/${candidateId}/outcome`,
    payload,
  );
  return data;
}

export async function getOutcomeApi(
  jobId: string,
  candidateId: string,
): Promise<CandidateJobOutcome | null> {
  const { data } = await apiClient.get<CandidateJobOutcome | null>(
    `/api/jobs/${jobId}/candidates/${candidateId}/outcome`,
  );
  return data;
}

export async function getCalibrationApi(
  jobId: string,
): Promise<CalibrationReport> {
  const { data } = await apiClient.get<CalibrationReport>(
    `/api/jobs/${jobId}/calibration`,
  );
  return data;
}

export async function getFunnelApi(jobId?: string): Promise<FunnelStats> {
  const { data } = await apiClient.get<FunnelStats>("/api/dashboard/funnel", {
    params: jobId ? { job_id: jobId } : undefined,
  });
  return data;
}
